ext/json/json_encoder.c

```diff
@@ -93,19 +93,26 @@ static inline void php_json_encode_double(smart_str *buf, double d, int options)
 }
 /* }}} */
 
-#define PHP_JSON_HASH_PROTECT_RECURSION(_tmp_ht) \
-	do { \
-		if (_tmp_ht) { \
-			GC_TRY_PROTECT_RECURSION(_tmp_ht); \
-		} \
-	} while (0)
-
-#define PHP_JSON_HASH_UNPROTECT_RECURSION(_tmp_ht) \
-	do { \
-		if (_tmp_ht) { \
-			GC_TRY_UNPROTECT_RECURSION(_tmp_ht); \
-		} \
-	} while (0)
+static inline zend_result php_json_protect_recursion(php_json_encoder *encoder, zend_refcounted *rc)
+{
+	if (GC_FLAGS(rc) & GC_IMMUTABLE) {
+		return SUCCESS;
+	}
+	if (zend_hash_index_add_empty_element(&encoder->recursive, (uintptr_t) rc)) {
+		GC_ADDREF(rc);
+		return SUCCESS;
+	}
+	return FAILURE;
+}
+
+static inline void php_json_unprotect_recursion(php_json_encoder *encoder, zend_refcounted *rc)
+{
+	if (GC_FLAGS(rc) & GC_IMMUTABLE) {
+		return;
+	}
+	GC_DELREF(rc);
```

Comment:
So, the original code was increasing the reference count of the properties table, and decreasing the reference count of the properties table, to prevent the property table from getting freed during iteration.

Now that we're referencing the object, your PR is calling GC_ADDREF and GC_DELREF to prevent the object from getting freed during iteration.

If this PR were to be reopened or a if someone were to base code on this in the future, would it need to check if GC_DELREF returns 0 and free the array/object in question, to avoid leaks if jsonSerialize removed the last reference to a value as a side effect? i.e. call rc_dtor_func if unexpectedly 0

`ZEND_API void ZEND_FASTCALL rc_dtor_func(zend_refcounted *p)`